### PR TITLE
Add IndexCache support for DeepSeek V4

### PR DESCRIPTION
**Pull request by andrewdoro**
## Summary

Adds opt-in IndexCache support for DeepSeek V4 CSA/C4 layers.

- Adds `index_topk_freq`, `index_topk_pattern`, and `index_topk_min_seq_len` config knobs. Defaults preserve existing behavior: `index_topk_freq=1`, no pattern, and a 75k-token runtime floor before reuse is active.
- Lets selected C4 layers skip their own indexer call and reuse the nearest preceding F layer's raw logical top-k indices.
- Rebuilds current-layer physical page indices from reused raw indices, the current page table, and current C4 sequence lengths instead of aliasing physical indices across layers.
- Keeps IndexCache disabled for NextN/EAGLE draft layers, non-C4 layers, and HiSparse coordinator mode.
- Adds optional trace instrumentation for CSA indexer, raw-to-page translation, core attention, FFN/MoE, and CUDA graph replay/fallback paths.
- Adds focused CPU tests for policy selection, pattern validation, sequence-length gating, HiSparse guards, raw-to-page translation, CUDA graph mode selection, and profiling fast-path behavior.

## Validation

Local checks on the current cleaned diff:

- `python -m pytest test/srt/cpu/test_deepseek_v4_index_cache.py -q` -> 115 passed
- focused `ruff check` -> passed
- focused `black --check` -> passed
- focused `python -m compileall` -> passed
- `git diff --check` -> passed

PR CI note: `lint` is green. The red PR Test Base/Extra jobs are draft gates; the workflow log reports `PR is draft. Blocking CI.`

## Scope

This PR intentionally contains only the upstream implementation and CPU coverage. Benchmark, Modal, profiling-driver, greedy-search, and eval-suite scripts are kept local and are not part of this PR.

## Remaining external validation

Before enabling broadly, run the separate local validation harness against real B200 endpoints:

- EAGLE-off profiler traces at long context, confirming indexer savings, raw-to-page overhead, sparse/core attention cost, FFN/MoE cost, and CUDA graph replay/fallback counts.
- Greedy training-free searched patterns for 1/2 and 1/4 retention. Do not use uniform 1/4 as a production candidate.
- Long-context quality suite: RULER, MRCR v2, GraphWalks, LongBench v2, and AA-LCR, plus AIME/GPQA/LiveCodeBench/IFBench for reasoning/general coverage.
- EAGLE-on perf/correctness after the base EAGLE-off path is validated.
- Context-gate sensitivity before lowering the default 75k threshold.

---
### CI States

Latest PR Test (Base): :x: [Run #26421024543](https://github.com/sgl-project/sglang/actions/runs/26421024543)
Latest PR Test (Extra): :x: [Run #26421024446](https://github.com/sgl-project/sglang/actions/runs/26421024446)

### CODE REVIEW

**Comment by andrewdoro** [comment]
Update after latest push:

- Fixed the CI lint failure by applying the repo pre-commit formatting hooks (`isort`, `black-jupyter`).
- Local checks: `ruff check` on touched DSV4 IndexCache files, `py_compile` on manual DSV4 scripts, and focused CPU tests all pass (`46 passed`).
- GitHub `lint` is now green on run 26371760005.

No new Modal/GPU validation was started in this pass.

**Comment by andrewdoro** [comment]
Update after latest push:

- Hardened the validation workflow so real quality evals require distinct searched `1/2` and searched `1/4` endpoints.
- The orchestration now always passes the `{pattern}` deployment template into greedy search, and validates that the template includes `{pattern}` before running.
- Local checks: focused IndexCache CPU tests pass (`53 passed`), plus `ruff`, `py_compile`, `isort`, and `black-jupyter` on touched files.
- GitHub `lint` is green on run 26372325307.

No Modal/GPU validation was started in this pass.

File: python/sglang/srt/models/deepseek_v4_index_cache.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+) -> None:
+    page_bits = (page_size - 1).bit_length() if page_size > 1 else 0
+    page_mask = page_size - 1
+    valid = raw_indices >= 0
+    if seq_lens.dim() > 1:
+        seq_lens = seq_lens.squeeze(-1)
+    valid = valid & (raw_indices < seq_lens.unsqueeze(1))
+    page_idx = torch.clamp(raw_indices >> page_bits, min=0)
+    offset_in_page = raw_indices & page_mask
+    physical_pages = torch.gather(page_table, dim=1, index=page_idx.long())
```
[the remark this answers is not in the document]
Comment (andrewdoro) [reply python/sglang/srt/models/deepseek_v4_index_cache.py]: Fixed in ffd097c0d. `page_idx` is now clamped to the valid page-table column range before `torch.gather`, and I added a regression test with an invalid far-out raw index to prove it masks to `-1` without crashing.

**Review by andrewdoro — COMMENTED** [review]
B200 EAGLE-off decode smoke result for the current gated `index_topk_freq=2` implementation.

Setup:
- baseline app: `index_topk_freq=1`
- IndexCache app: `index_topk_freq=2`
- `index_topk_min_seq_len=75000`
- EAGLE/spec decode off on both endpoints, confirmed through `/server_info`
- B200:4 Modal apps, same branch/build
- smoke size: 3 requests per shape, `max_tokens=128`
- request errors: 0

IndexCache vs baseline aggregate completion throughput:

```text
prompt   c1      c2      c4
75k      +30%    +32%    +50%
100k     +35%    +36%    +21%
128k     +36%    +39%    +45%
178k     +41%    +36%    +36%
200k     +44%    +50%    +50%
300k     +47%    +50%    +61%
```

P50 request wall time improved on every measured bucket, roughly -17% to -38%.

Caveat: this is a directional smoke, not the final production estimate. The earlier 200-request full sweep was too slow at serial long-context settings, so the next validation should be a focused larger-N run on the crossover/long-tail buckets, especially 178k/200k/300k at concurrency 1/2/4, plus quality eval before marking the optimization shippable.

**Review by andrewdoro — COMMENTED** [review]
Follow-up prod-traffic pressure bench using Anara's `bench_anara_traffic.py` 

Setup:
- baseline app: `index_topk_freq=1`
- IndexCache app: `index_topk_freq=2`
- EAGLE/spec decode off on both endpoints
- `--multi-turn --prod-dist --output-dist`
- `--concurrency 12 --chat-pool-size 80 --duration 90 --trials 1 --pool-seed 26230`
- one B200:4 Modal app per arm

Results:

```text
metric              baseline   freq2     delta
errors              0          0         same
n_ok                26         39        +50%
actual rps          0.289      0.433     +50%
p50 wall            35.3s      26.4s     -25%
p95 wall            105.1s     71.7s     -32%
p99 wall            124.9s     104.4s    -16%
mean wall           45.0s      31.4s     -30%
agg prompt tok/s    41.1k      56.9k     +39%
```

By-kind p50 wall:

```text
cold_long   35.3s -> 23.4s   -34%
multi_turn  40.2s -> 20.7s   -49%
mid         54.2s -> 28.8s   -47%
```

Interpretation: c12 is stable and positive for freq2: zero errors on both arms, higher completed throughput, lower p50/p95/p99 wall time, and +39% aggregate prompt-token throughput. This contrasts with the c50 pressure run, which overloaded both arms and produced high client timeout rates, so c50 should not be treated as a clean capacity estimate.

Artifacts:
- `/tmp/bench_indexcache_pressure_c12/indexcache_baseline_pressure_c12_trial0.json`
- `/tmp/bench_indexcache_pressure_c12/indexcache_freq2_pressure_c12_trial0.json`

**Review by andrewdoro — COMMENTED** [review]
Longer prod-traffic pressure bench at the same stable concurrency (`c12`) to reduce small-N noise.

Setup:
- same Anara `bench_anara_traffic.py` pressure mode from PR #3586
- baseline app: `index_topk_freq=1`
- IndexCache app: `index_topk_freq=2`
- EAGLE/spec decode off on both endpoints
- `--multi-turn --prod-dist --output-dist`
- `--concurrency 12 --chat-pool-size 80 --duration 240 --trials 1 --pool-seed 26231`
- one B200:4 Modal app per arm

Results:

```text
metric              baseline   freq2     delta
errors              0          0         same
n_ok                66         94        +42%
actual rps          0.275      0.392     +42%
p50 wall            34.7s      16.8s     -51%
p95 wall            118.0s     91.0s     -23%
p99 wall            130.9s     123.3s    -6%
mean wall           42.5s      29.1s     -32%
agg prompt tok/s    38.5k      58.6k     +52%
total prompt toks   9.24M      14.07M    +52%
```

By-kind p50 wall:

```text
cold_long   37.1s -> 15.2s   -59%
multi_turn  24.8s -> 19.7s   -21%
mid         16.5s -> 17.1s   +4%
```

Both arms were clean at c12 with zero timeouts. The stronger 240s run confirms the earlier 90s c12 signal: freq2 completes materially more production-shaped traffic and improves p50/p95 wall time under stable pressure. The c50 run should still be treated as an overload probe only, not a capacity estimate.

Artifacts:
- `/tmp/bench_indexcache_pressure_c12_240s/indexcache_baseline_pressure_c12_240s_trial0.json`
- `/tmp/bench_indexcache_pressure_c12_240s/indexcache_freq2_pressure_c12_240s_trial0.json`

**Review by andrewdoro — COMMENTED** [review]
Additional production-shaped pressure bench at higher stable load: `concurrency=20`, `duration=340s`.

Setup:
- same Anara `bench_anara_traffic.py` pressure mode from PR #3586
- baseline app: `index_topk_freq=1`
- IndexCache app: `index_topk_freq=2`
- EAGLE/spec decode off on both endpoints
- `--multi-turn --prod-dist --output-dist`
- `--concurrency 20 --chat-pool-size 80 --duration 340 --trials 1 --pool-seed 26232`
- one B200:4 Modal app per arm

Results:

```text
metric              baseline   freq2     delta
n_ok                126        167       +32.5%
errors              1          0         improved
actual rps          0.371      0.491     +32.5%
p50 wall            39.6s      25.2s     -36.2%
p95 wall            110.5s     99.1s     -10.3%
p99 wall            211.7s     143.6s    -32.1%
mean wall           50.7s      35.9s     -29.1%
agg prompt tok/s    48.2k      64.4k     +33.8%
total prompt toks   16.37M     21.90M    +33.8%
```

By-kind p50 wall:

```text
cold_long   35.4s -> 25.4s   -28.3%
multi_turn  41.0s -> 24.3s   -40.6%
mid         31.8s -> 26.4s   -16.8%
```

This sits between the clean c12 runs and the overloaded c50 probe. At c20, freq2 still completes more production-shaped traffic with lower p50/mean/p99 wall time and no errors. p95 improvement is smaller (-10%), so the knee is starting to show, but this is still a positive stable-load result.

Artifacts:
- `/tmp/bench_indexcache_pressure_c20_340s/indexcache_baseline_pressure_c20_340s_trial0.json`
- `/tmp/bench_indexcache_pressure_c20_340s/indexcache_freq2_pressure_c20_340s_trial0.json`

**Comment by andrewdoro** [comment]
EAGLE-on verification after the target-verify gate in `e87e13000`.

Change tested:
- keep EAGLE enabled
- `index_topk_freq=2`
- `index_topk_min_seq_len=75000`
- disable IndexCache reuse/return during `ForwardMode.TARGET_VERIFY`
- keep EAGLE draft CUDA graph runners out of IndexCache graph variants

Setup:
- exact prod-shape `2 x B200:4` Modal app per arm, via Anara `deploy_prod.py`
- EAGLE on, CP on
- `MEM_FRACTION=0.74`
- deterministic prod-traffic bench: `--multi-turn --prod-dist --output-dist --concurrency 20 --chat-pool-size 80 --duration 340 --trials 1 --pool-seed 26232`
- baseline: `index_topk_freq=1`
- candidate: `index_topk_freq=2`
- both arms returned sane `OK` completions before the bench
- request errors: 0 on both arms

Results:

```text
metric              baseline   freq2+verify-gate   delta
n_ok                183        219                 +19.7%
errors              0          0                   same
actual rps          0.538      0.644               +19.7%
p50 wall            21.8s      21.1s               -3.3%
p95 wall            105.7s     69.6s               -34.2%
p99 wall            171.4s     116.0s              -32.4%
mean wall           33.5s      27.9s               -16.6%
agg prompt tok/s    70.6k      84.0k               +19.0%
total prompt toks   23.99M     28.56M              +19.0%
```

By-kind p95:

```text
cold_long   101.25s -> 67.93s   -32.9%
multi_turn  105.73s -> 69.61s   -34.2%
mid         105.22s -> 56.04s   -46.7%
```

Interpretation: the earlier EAGLE-on regression was not inherent to IndexCache. It was from reusing stale CSA top-k during EAGLE target verification. Recomputing fresh top-k for verify preserves the EAGLE path and makes `freq=2` positive in the exact 2x4xB200 shape: +19% prompt throughput and materially better tail latency.

Artifacts:
- `/tmp/bench_eagle_verifygate_2x_c20_340s/eagle_vg2x_baseline_c20_340s_trial0.json`
- `/tmp/bench_eagle_verifygate_2x_c20_340s/eagle_vg2x_freq2_c20_340s_trial0.json`

Both Modal test apps were stopped after the run.
